src/finch/interface/tensor.py

```diff
@@ -20,15 +20,283 @@ def to_numpy(self):
+    @abstractmethod
+    def __ror__(self, other):
+        """Define right bitwise OR for tensors."""
+        lazy.compute(lazy.lazy(self).__ror__(other))
+    
+    @abstractmethod
+    def __rxor__(self, other):
+        """Define right bitwise XOR for tensors."""
+        lazy.compute(lazy.lazy(self).__rxor__(other))
+
+    def __iadd__(self, other):
+        """Define in-place addition for tensors."""
+        lazy.compute(lazy.lazy(self).__iadd__(other))
+    
+    def __isub__(self, other):
+        """Define in-place subtraction for tensors."""
+        lazy.compute(lazy.lazy(self).__isub__(other))
+    
+    def __imul__(self, other):
+        """Define in-place multiplication for tensors."""
+        lazy.compute(lazy.lazy(self).__imul__(other))
+
+    def __imatmul__(self, other):
+        """Define in-place matrix multiplication for tensors."""
+        lazy.compute(lazy.lazy(self).__imatmul__(other))
+
+    def __itruediv__(self, other):
+        """Define in-place true division for tensors."""
+        lazy.compute(lazy.lazy(self).__itruediv__(other))
+
+    def __ifloordiv__(self, other):
+        """Define in-place floor division for tensors."""
+        lazy.compute(lazy.lazy(self).__ifloordiv__(other))
+
+    def __imod__(self, other):
+        """Define in-place modulo for tensors."""
+        lazy.compute(lazy.lazy(self).__imod__(other))
+    
+    def __idivmod__(self, other):
+        """Define in-place division modulo for tensors."""
+        lazy.compute(lazy.lazy(self).__idivmod__(other))
+
+    def __ipow__(self, other):
+        """Define in-place power for tensors."""
+        lazy.compute(lazy.lazy(self).__ipow__(other))
+
+    def __ilshift__(self, other):
+        """Define in-place left shift for tensors."""
+        lazy.compute(lazy.lazy(self).__ilshift__(other))
+    
+    def __irshift__(self, other):
+        """Define in-place right shift for tensors."""
+        lazy.compute(lazy.lazy(self).__irshift__(other))
+
+    def __iand__(self, other):
+        """Define in-place bitwise AND for tensors."""
+        lazy.compute(lazy.lazy(self).__iand__(other))
+    
+    def __ior__(self, other):
+        """Define in-place bitwise OR for tensors."""
+        lazy.compute(lazy.lazy(self).__ior__(other))
+
+    def __ixor__(self, other):
+        """Define in-place bitwise XOR for tensors."""
+        lazy.compute(lazy.lazy(self).__ixor__(other))
+
+    def __neg__(self):
+        """Define negation for tensors."""
+        lazy.compute(lazy.lazy(self).__neg__())
+    
+    def __pos__(self):
+        """Define unary plus for tensors."""
+        lazy.compute(lazy.lazy(self).__pos__())
+    
+    def __abs__(self):
+        """Define absolute value for tensors."""
+        lazy.compute(lazy.lazy(self).__abs__())
+
+    def __invert__(self):
+        """Define bitwise NOT for tensors."""
+        lazy.compute(lazy.lazy(self).__invert__())
+
+    #TODO: should these raise an error? There are few others like that
```

Comment:
These should succeed only for 0-dimensional arrays and/or scalars, provided they can be safely coerced. i.e. if the array is floating point, only `__float__` and `__complex__` should succeed.

See the type promotion rules here: https://data-apis.org/array-api/latest/API_specification/type_promotion.html

Reply:
See also, e.g. https://data-apis.org/array-api/latest/API_specification/generated/array_api.array.__int__.html#array_api.array.__int__